operators/pkg/instautoctrl/terminator.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	// Check the selector label, in order to know whether to perform or not reconciliation.
+	if proceed, err := utils.CheckSelectorLabel(ctx, r.Client, instance.GetNamespace(), r.NamespaceWhitelist.MatchLabels); !proceed {
+		// If there was an error while checking, show the error and try again.
+		if err != nil {
+			log.Error(err, "failed checking selector labels")
+		}
+		return ctrl.Result{}, err
+	}
+
+	requeueResultOnError := ctrl.Result{RequeueAfter: InstanceStatusCheckRequeueTimeOnError}
+
+	// Check if the instance has to be terminated.
+	terminate, err := r.CheckInstanceTermination(ctx, &instance)
+	if err != nil {
+		log.Error(err, "failed checking instance termination")
+		return requeueResultOnError, err
+	}
+
+	if terminate {
+		if err := r.TerminateInstance(ctx, &instance); err != nil {
+			log.Error(err, "failed terminating instance")
+			return requeueResultOnError, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("requeueing instance")
+	return ctrl.Result{RequeueAfter: InstanceStatusCheckRequeueTime}, nil
+}
+
+// CheckInstanceTermination checks if the Instance has to be terminated.
+func (r *InstanceTerminationReconciler) CheckInstanceTermination(ctx context.Context, instance *clv1alpha2.Instance) (bool, error) {
+	if instance.Spec.CustomizationUrls == nil {
+		return false, errors.New("customization urls field is not set for Instance")
+	}
+
+	statusCheckURL := instance.Spec.CustomizationUrls.StatusCheck
+	if statusCheckURL == "" {
+		return false, errors.New("status check url field is not set for Instance")
+	}
+
+	log := ctrl.LoggerFrom(ctx).WithName("statusCheck")
+	log.Info("performing instance status check")
+
+	statusCheckReponse := StatusCheckResponse{}
+	statusCode, err := utils.HTTPGetJSONIntoStruct(ctx, statusCheckURL, &statusCheckReponse)
+	if statusCode != http.StatusNotFound && err != nil {
+		return false, err
+	}
+
+	status := instance.Status
+	status.Automation.LastCheckTime = metav1.Now()
+	if statusCode == http.StatusOK {
+		status.Automation.TerminationTime = metav1.Time{Time: statusCheckReponse.Deadline}
+	}
+
+	if err := r.PatchStatus(ctx, instance, &status); err != nil {
+		log.Error(err, "failed patching instance status")
+		return false, err
+	}
+
+	switch statusCode {
+	case http.StatusOK:
+		log.Info("termination not required")
+		return false, nil
+	case http.StatusNotFound:
+		log.Info("termination required")
+		return true, nil
+	default:
+		return false, fmt.Errorf("failed: unexpected status code %d, retrieved error='%s'", statusCode, statusCheckReponse.Error)
+	}
+}
+
+// TerminateInstance terminates the Instance.
+func (r *InstanceTerminationReconciler) TerminateInstance(ctx context.Context, instance *clv1alpha2.Instance) error {
+	log := ctrl.LoggerFrom(ctx).WithName("termination")
+	log.Info("terminating instance")
+
+	instance = &clv1alpha2.Instance{ObjectMeta: forge.ObjectMeta(instance)}
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, instance, func() error {
```

Comment:
Better to avoid CreateOrUpdate here, and use Update directly since:
* The instance exists for sure (and shouldn't be recreated if deleted in the meantime)
* The get operation carried out by the CreateOrUpdate function might cause a conflict, since you are possibly reading old data.

Reply:
Isn't it better to use a Patch then? Wouldn't an Update fail in case anyone else changed the Instance in the meantime?

Reply:
Would personally avoid using patch whenever possible, as it is associated with higher complexity and possible pitfalls (would be better to remove that from the status update too). If everything is written correctly, you should not incur in self-update conflicts, and the rate of other updates is definitely low, so that should not lead to problems. Still, in case of conflicts it is better to retry immediately, instead of waiting for the 5 seconds requeue time.

Reply:
Would personally avoid using patch whenever possible, as it is associated with higher complexity and possible pitfalls (would be better to remove that from the status update too). If everything is written correctly, you should not incur in self-update conflicts, and the rate of other updates is definitely low, so that should not lead to problems. Still, in case of conflicts it is better to retry immediately, instead of waiting for the 5 seconds requeue time.

Reply:
Don't know if I did it in the right way

Reply:
That makes sense to me. (Could be optimized checking if an update is actually necessary, but that might lead to possible inconsistencies, in case the status is updated at the same time. BTW, that might also be a reason for the bug in the instance controller if still present)

Reply:
The current setup seems to be functional. Is there anything else to be done?

Reply:
I would say not